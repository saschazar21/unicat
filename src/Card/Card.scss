@import '../__styles__/variables';

.body {
  text-align: center;
  max-height: 100%;
  overflow-x: hidden;
  overflow-y: auto;

  .picture + & {
    margin-top: spacing(xl);
  }

  &:not(:last-child) {
    padding-bottom: spacing(xl);

    &.light {
      padding-bottom: spacing();
    }
  }
}

.cta {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  padding-top: spacing(xl);

  &.light {
    padding-top: spacing();
  }

  .body + & {
    border-top: borderWidth() dashed $borderColor;
  }

  @include media-query-up(m) {
    grid-template-columns: 1fr 1fr;
    grid-column-gap: spacing();
  }
}

.wrapper {
  position: relative;
  background-color: color(white, false, $systemColors);
  color: color(default);
  border: border();
  border-radius: borderRadius();
  padding: spacing(xl);
  margin: spacing(m);
  width: 100%;
  box-shadow: boxShadow(m);
  overflow: hidden;
}

.dark {
  background-color: color(default);
  color: color(white, false, $systemColors);
}

.light {
  padding: spacing();
  margin: spacing(s);
  box-shadow: boxShadow(xs);
}

.controls {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  flex-wrap: nowrap;
  margin-bottom: spacing(s);
  min-width: 0;
  max-width: 100%;
}

.picture {
  display: block;
  position: relative;
  margin: -#{spacing(xl)};

  .light > & {
    margin: -#{spacing()};
  }

  &::after {
    content: '';
    position: absolute;
    background: linear-gradient(
      180deg,
      RGBA(color(white, false, $systemColors, true), 0),
      RGBA(color(white, false, $systemColors, true), 0.5) 70%,
      RGBA(color(white, false, $systemColors, true), 0.7) 70%,
      color(white, false, $systemColors) 90%
    );
    top: 50%;
    right: 0;
    bottom: 0;
    left: 0;
  }

  .dark > &::after {
    background: linear-gradient(
      180deg,
      RGBA(color(default, false, $colors, true), 0),
      RGBA(color(default, false, $colors, true), 0.5) 50%,
      RGBA(color(default, false, $colors, true), 0.7) 70%,
      color(default) 90%
    );
  }
}
