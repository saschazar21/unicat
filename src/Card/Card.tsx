import React, { Component, ReactNode, ReactElement } from 'react';
import classnames from 'classnames';
import { CloseIcon } from '@saschazar/unicat-icons';

import { SmallVariant } from '../__types__/global';
import Button from '../Button';

import styles from './Card.scss';
import IconButton from '../IconButton';

export interface CardProps {
  children: ReactNode[] | ReactNode;
  cta?: ReactElement<Button>;
  className?: string;
  dark?: boolean;
  image?: ReactNode;
  modal?: boolean;
  onClose?: (event?: React.SyntheticEvent) => void;
  variant?: SmallVariant;
}

export default class Card extends Component<CardProps> {
  static defaultProps = {
    variant: 'default',
  };

  public renderCta = () => {
    const { cta } = this.props;

    return <div className={styles.cta}>{cta}</div>;
  };

  public renderControls = () => {
    const { onClose } = this.props;

    return (
      <div className={styles.controls}>
        {onClose && (
          <IconButton
            variant="light"
            name="close"
            icon={<CloseIcon />}
            onClick={onClose}
          />
        )}
      </div>
    );
  };

  public render() {
    const { variant: defaultVariant } = Card.defaultProps;
    const {
      children,
      className: customClassName,
      cta,
      dark,
      image,
      modal,
      onClose,
      variant = defaultVariant,
    } = this.props;

    const aria = modal ? { role: 'dialog', 'aria-modal': true } : {};

    const className = classnames(
      styles.wrapper,
      { [styles[variant]]: variant !== defaultVariant },
      { [styles.dark]: dark || image },
      customClassName
    );

    const bodyClassName = classnames(styles.body);

    return (
      <aside className={className} {...aria}>
        {onClose && this.renderControls()}
        {image && <picture className={styles.picture}>{image}</picture>}
        <div className={bodyClassName}>{children}</div>
        {cta && this.renderCta()}
      </aside>
    );
  }
}
