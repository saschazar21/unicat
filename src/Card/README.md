Renders a Box with a `box-shadow`.

```jsx
import Heading from '../Heading/Heading.tsx';

<Card>
  <Heading level="h2">Headline</Heading>
  <span>A default Card</span>
</Card>;
```

If there's a special need for a smaller use case, the `light` variant prop type may also be used.

```jsx
import Heading from '../Heading/Heading.tsx';

<Card variant="light">
  <Heading level="h2">Headline</Heading>
  <span>A smaller version of the Card.</span>
</Card>;
```
