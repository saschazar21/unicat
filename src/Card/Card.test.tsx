import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Card from './Card';

describe('Card', () => {
  it('renders', () => {
    const wrapper = shallow(<Card><span>Test</span></Card>);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('span')).toHaveLength(1);
  });

  it('renders a light variant', () => {
    const wrapper = shallow(<Card variant="light">Test</Card>);

    expect(toJson(wrapper)).toMatchSnapshot();

    expect(wrapper.find('.light')).toHaveLength(1);
  });
});