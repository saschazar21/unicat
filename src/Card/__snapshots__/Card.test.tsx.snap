// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders 1`] = `
<aside
  className="wrapper"
>
  <span>
    Test
  </span>
</aside>
`;

exports[`Card renders a light variant 1`] = `
<aside
  className="wrapper light"
>
  Test
</aside>
`;
